### PR TITLE
Introduce DeclExceptionMsg.

### PR DESCRIPTION
**Pull request by bangerth**
In response to the discussion in #610

### CODE REVIEW

File: include/deal.II/base/exceptions.h [hunk 1]
```diff
@@ -377,6 +377,21 @@ namespace deal_II_exceptions
 
 
 /**
+ * Declare an exception class derived from ExceptionBase with that takes no
+ * additional arguments but still has a message that it prints.
+ *
+ * @ingroup Exceptions
+ */
+#define DeclExceptionMsg(Exception, outsequence)                    \
+  class Exception1 : public dealii::ExceptionBase {                       \
```
Comment (tjhei) [inline include/deal.II/base/exceptions.h]: The name of the class needs to be the same as the first argument.

Comment (bangerth) [inline include/deal.II/base/exceptions.h]: Oh, woops...

**Comment by tjhei** [comment]
I don't understand. This doesn't allow you to provide an individual message at the location of the Assert (see my suggestion at #610) so you are forced to use ExcMessage if you have something to say, right?

**Comment by bangerth** [comment]
Correct. I'm thinking of exceptions that are thrown in a number of places (think, the old ExcNumberNotFinite or ExcTriangulationNotEmpty) where there is nothing really good that can be gleaned from the context -- so there are no runtime arguments. In cases like this, providing a message at the place where it happens makes no sense because it has to be repeated in multiple places.

You are talking about an exception that has the same name but would produce a different message in different contexts. Where would that happen?

**Comment by tjhei** [comment]
Did you see my example at the bottom of #610 ? My suggestion would do both: 1) allow you to give a default message for something like ```ExcInternalError``` and 2) optionally give a specific message.

(in my example I am replacing the default message but we can talk about printing both though)

**Comment by bangerth** [comment]
Ah, I missed that. Yes, nice trick. Will take care of this when I come back this afternoon, it clearly looks like the right way to go.

**Comment by bangerth** [comment]
Like this?

**Comment by tjhei** [comment]
Like it! 

1. Did you test that overriding the default message works?
2. Does it make sense to always attach the default message below the custom message? (No idea)

**Comment by bangerth** [comment]
I verified that both default text and replacement text are correctly printed. (We can't do this in the testsuite because -- maybe that's a bug -- when we tell `Assert` to throw an exception instead of aborting the program, it forgets about the additional information in the exception. But I've verified this by hand.)

As for your point 2 -- I don't know. I don't have a use case where I'd like to replace the text right now. I'd say we make that decision when we have a use case. No point engineering a solution without knowing what we need.

**Comment by tjhei** [comment]
:+1:

**Pull request merged**